packages/wonder-blocks-core/util/get-element-intersection.js

```diff
@@ -0,0 +1,130 @@
+// @flow
+import enumerateScrollAncestors from "./enumerate-scroll-ancestors.js";
+
+/**
+ * Indicates the intersection state of an item on a single axis.
+ */
+export type AxisIntersection =
+    /**
+     * The item is partly or fully within the bounds.
+     */
+    | "within"
+
+    /**
+     * The item is outside the starting bounds.
+     */
+    | "before"
+
+    /**
+     * The item is outside the ending bounds.
+     */
+    | "after";
+
+/**
+ * Indicates the visibility of an item on the horizontal and vertical axes.
+ */
+export type Intersection = {|
+    /**
+     * The intersection on the horizontal axis.
+     */
+    horizontal: ?AxisIntersection,
+
+    /**
+     * The intersection on the vertical axis.
+     */
+    vertical: ?AxisIntersection,
+|};
+
+const UndeterminedIntersection: Intersection = {
+    horizontal: null,
+    vertical: null,
+};
```

Comment:
I wonder if it's worth `freeze`-ing or just returning a new value each time.  Both have their own runtime costs, but returning a shared object feels dangerous.

Reply:
Yeah, I wondered about that myself. Freezing sounds like a great idea, but we could also change this to be a function that returns the result, so we still encapsulate in one place.

Reply:
Yeah, let's go with that.  We should do the same for `FullIntersection`.